### PR TITLE
Pytorch integration

### PR DESCRIPTION
**Pull request by orausch**
cc @Shigangli

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,6 @@ install:
   - make install
   - pip install codecov pytest-cov
 script:
-  - make test
+  - travis_wait 30 make test
```
Comment (tbennun) [inline .travis.yml:12]: Could you explain this?

File: daceml/onnx/nodes/onnx_op.py [hunk 2]
```diff
@@ -337,6 +338,56 @@ def validate(self, sdfg: SDFG, state: SDFGState):
...
+        # which fails in the case of ONNX ops
+        def _register_implementation(self, name, transformation_type):
+            self.implementations[name] = transformation_type
+            match_node_name = "__" + transformation_type.__name__
+            if (hasattr(transformation_type, "_match_node")
+                    and transformation_type._match_node != match_node_name):
+                raise ValueError(
+                    "Transformation " + transformation_type.__name__ +
+                    " is already registered with a different library node.")
+            transformation_type._match_node = self
```
Comment (tbennun) [inline daceml/onnx/nodes/onnx_op.py:354]: Is this something we can merge into dace on its own? Why is there a need to instantiate the node in the first place?

File: daceml/pytorch/module.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from functools import wraps
+
+import torch.nn as nn
+import onnx
+
+from daceml.onnx import ONNXModel
+from daceml.onnx.shape_inference import infer_shapes
+
+
+class DACEModule(nn.Module):
```
Comment (tbennun) [inline daceml/pytorch/module.py:17]: Naming convention is a bit weird. Please use either `DaceModule` (standard Python) or `DaCeModule`.

File: daceml/pytorch/module.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        super(DACEModule, self).__init__()
+
+        self.model = model
+        self.sdfg = None
+        if dummy_inputs is not None:
+            self.dace_model = self.initialize_sdfg(dummy_inputs)
+
+    def initialize_sdfg(self, dummy_inputs) -> ONNXModel:
+        export_fd, export_name = tempfile.mkstemp(suffix=".onnx")
+        infer_fd, infer_name = tempfile.mkstemp(suffix=".onnx")
```
Comment (tbennun) [inline daceml/pytorch/module.py:28]: could use `with` statements instead of the `os.close`...

File: tests/test_bert/test_bert.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import os
```
Comment (tbennun) [inline tests/test_bert/test_bert.py:1]: Unrelated test but ok

**Review by tbennun — CHANGES_REQUESTED** [review]

**Comment by tbennun** [comment]
@orausch @Shigangli Also, since we have infer_shapes in our code, couldn't we make another function that doesn't call `onnx.load` and simply uses our existing loaded model?

File: daceml/onnx/nodes/onnx_op.py (hunk 2 above)
Comment (orausch) [reply daceml/onnx/nodes/onnx_op.py:354]: Here is the original method: https://github.com/spcl/dace/blob/master/dace/sdfg/nodes.py#L1187. I'm not sure how `_match_node` is used internally though

File: daceml/pytorch/module.py (hunk 4 above)
Comment (orausch) [reply daceml/pytorch/module.py]: I looked at this for a while and I'm not really sure how; `TemporaryFile` does have a context manager, but it returns a file object, which doesn't work with `onnx.load`.

File: .travis.yml (hunk 1 above)
Comment (orausch) [reply .travis.yml:13]: Added a comment

File: daceml/pytorch/module.py (hunk 4 above)
Comment (tbennun) [reply daceml/pytorch/module.py]: you can use the created object's `name` field to get the file path iirc

File: daceml/onnx/nodes/onnx_op.py (hunk 2 above)
Comment (tbennun) [reply daceml/onnx/nodes/onnx_op.py:354]: It's used as part of the expand transformation to create the subgraph. As long as it's hashable and unique to the transformation object it should be fine.

**Review by tbennun — APPROVED** [review]

**Pull request merged**